File: articles/lagrangian-and-hamiltonian-mechanics/fortran/hamiltonian_table.f90
! Hamiltonian Table
!
! This Fortran workflow computes the simple pendulum Hamiltonian:
!
!   H = p^2/(2 m l^2) + m g l(1 - cos(theta))

program hamiltonian_table
  implicit none

  integer :: i, j
  real(8), parameter :: mass_kg = 1.0d0
  real(8), parameter :: length_m = 1.0d0
  real(8), parameter :: g = 9.80665d0
  real(8), parameter :: pi_value = 3.141592653589793d0
  real(8), dimension(5) :: theta_values
  real(8), dimension(5) :: p_values
  real(8) :: theta
  real(8) :: p
  real(8) :: h

  theta_values = (/-pi_value, -pi_value/2.0d0, 0.0d0, pi_value/2.0d0, pi_value/)
  p_values = (/-4.0d0, -2.0d0, 0.0d0, 2.0d0, 4.0d0/)

  print *, "theta_rad,p_theta_kg_m2_per_s,hamiltonian_j"

  do i = 1, size(theta_values)
     do j = 1, size(p_values)
        theta = theta_values(i)
        p = p_values(j)
        h = p**2 / (2.0d0 * mass_kg * length_m**2) + &
            mass_kg * g * length_m * (1.0d0 - cos(theta))

        print '(F14.8,A,F14.8,A,F18.8)', theta, ",", p, ",", h
     end do
  end do

end program hamiltonian_table

File: articles/lagrangian-and-hamiltonian-mechanics/c/hamiltonian_table.c
/*
Hamiltonian Table

This C workflow computes the simple pendulum Hamiltonian:

    H = p^2/(2 m l^2) + m g l(1 - cos(theta))
*/

#include <math.h>
#include <stdio.h>

double hamiltonian(double theta, double p, double mass, double length, double g) {
    return p * p / (2.0 * mass * length * length)
        + mass * g * length * (1.0 - cos(theta));
}

int main(void) {
    const double pi = 3.14159265358979323846;
    const double mass = 1.0;
    const double length = 1.0;
    const double g = 9.80665;

    double theta_values[] = {-pi, -pi / 2.0, 0.0, pi / 2.0, pi};
    double p_values[] = {-4.0, -2.0, 0.0, 2.0, 4.0};

    int n_theta = sizeof(theta_values) / sizeof(theta_values[0]);
    int n_p = sizeof(p_values) / sizeof(p_values[0]);

    printf("theta_rad,p_theta_kg_m2_per_s,hamiltonian_j\n");

    for (int i = 0; i < n_theta; i++) {
        for (int j = 0; j < n_p; j++) {
            printf(
                "%.8f,%.8f,%.8f\n",
                theta_values[i],
                p_values[j],
                hamiltonian(theta_values[i], p_values[j], mass, length, g)
            );
        }
    }

    return 0;
}
